inference-engine/src/transformations/include/ngraph_ops/type_relaxed.hpp

```diff
@@ -0,0 +1,208 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+#include <vector>
+#include <algorithm>
+#include <string>
+
+#include <transformations_visibility.hpp>
+
+#include "ngraph/op/op.hpp"
+
+namespace ngraph {
+namespace op {
+
+/// A base class for templated TypeRelaxed that maintains overriden input types and outtypes for an operations.
+class TRANSFORMATIONS_API TypeRelaxedBase {
+public:
+    virtual ~TypeRelaxedBase();
+
+    TypeRelaxedBase(
+            const std::vector<element::Type>& _input_data_types = {},
+            const std::vector<element::Type>& _output_data_types = {}) :
+            m_input_data_types(_input_data_types),
+            m_output_data_types(_output_data_types) {
+    }
+
+    /// \return Data type that will be set for output with a given index outputIndex.
+    /// If output with a specified index outputIndex hasn't been set before, element::undefined will returned.
+    /// Undefined means no type override happens for a given outputIndex and it will deduced as original
+    /// operation defineds in its infer function.
+    const element::Type& get_overriden_output_type(size_t outputIndex = 0) const {
+        if (outputIndex >= m_output_data_types.size()) {
+            return element::undefined;
```

Comment:
May be its better to throw ngrpaph_error here?

Reply:
No, the filling of overridden outputs/orig inputs are optional, so this class doesn't really know the real number of items. They all are undefined.